types/recommendation.go

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2019 The MayaData Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CStorPoolClusterRecommendation is a kubernetes custom
+// resource that represents recommended configurations
+// for one given input/requirement.
+type CStorPoolClusterRecommendation struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec           CStorPoolClusterRecommendationSpec        `json:"spec"`
+	RequestDetails CStorPoolClusterRecommendationRequestSpec `json:"requestDetails"`
+}
+
+// CStorPoolClusterRecommendationSpec contains recommended
+// block devices topology (blockdevices group by node)
+type CStorPoolClusterRecommendationSpec struct {
+	BlockDeviceTopology []BlockDeviceTopology `json:"blockDeviceTopology"`
```

Comment:
Feel the need to add the estimated capacity of each nstance here. Else, the server or UI will have to calculate the same.
Eg. User made a request for 80GB capacity, but as per the available devices, it can be 100GB. In this case, capacity in request spec in 80 and response spec is 100. Also better to keep this per instance config.

Maybe, introduce a new structure as Instance config which has the following:
Instance config {
  Node reference
  Capacity
  BlockDeviceTopology
}
and have an array of InstanceConfig in the CSPCRecommendationSpec

Reply:
cc @AmitKumarDas @vishnuitta

Reply:
I could not understand this fully. I have an impression to have all pools with all disks having same capacity.
We should have some way to approximate the device e.g. 80GiB can have 80GiB till 100GiB disks only. I guess resource.Quantity allows this as a %age field. Not too sure.

Reply:
Agreed, all pools can have same capacity. But this capacity needs to be stored as part of recommendation spec so that server/UI need not have to calculate them again.